### PR TITLE
Mutex guard lock ref recovery methods

### PR DESCRIPTION
**Pull request by ijackson**
## Motivation

There are various situations where it is useful to be able to recover `&Mutex` from a `MutexGuard`.  One of them is condition variables (see #3892 and my new [async-condvar-fair](https://crates.io/crates/async-condvar-fair) crate) but there are other possibilities.

## Solution

Broadly, for the applicable guard types, and where possible:
```
impl Guard<'l, T> {
    pub fn mutex(this: &Self) -> &'l Mutex<T> { self.lock }
}
```
(Mutatis mutandis for the various mutex types - varies eg for `Owned` and `RwLock`.)

In #3741 a method was proposed to *convert* the guard into the mutex reference.  Obtaining a borrow is strictly more general, and is what is implemented in `parking_lot` and `smol` for their mutex types.

## Open questions

 * Are these the best names?  `parking_lot` uses `fn mutex()` for its mutex type but does not provide these facilities for its rwlocks.  `smol` uses <strike>`lock()`</strike> `source()` which is more general, leading to less variation between `Mutex` and `RwLock`.

 * <strike>Was I right to use `#[tokio::main] async fn main(){...}` for my doctests?  The contribution guide suggests something involving `block_on` but the other tests in `mutex.rs` used `#[tokio::main]` so I chose that.</strike>

 * <strike>I wasn't sure whether to `#` out the `use`s at the top of the doctests.  I find them noisy here - in this particular case the intended types are extremely obvious.</strike>

 * <strike>I have skipped `RwLockReadGuard` and `RwLockWriteGuard`.  This is not because I think they wouldn't be useful, but those structs contain references to pieces of the rwlock, not the whole rwlock, and I wasn't sure how to recover a suitable reference (or whether that's even possible).</strike>  The (straightforward portions of) the corresponding feature for `RwLock` is now #3930.

### CODE REVIEW

**Comment by Darksonn** [comment]
I think having a method like this makes sense. Some comments:

 1. I prefer to add the method to all `RwLock` guards, or to none. It should be possible to add it to the remaining guards by changing the `&'a Semaphore` field to `&'a RwLock`.
 2. We should not call the method `lock()`. That is confusing since we already have a [`Mutex::lock`][1] method, and that method does something different.
 3. Using `#[tokio::main] async fn main(){...}` for the doctests is fine, though all the other doctests put the example directly in `main` rather than define a new method. I think I'm ok either way, but putting it directly in `main` reduces the amount of hidden code.
 4. I'd like to include the imports in the doctests. Without them, the example would not compile if pasted into the playground.
 5. Please run the doctests through rustfmt for more consistent formatting with the rest of the tests.
 6. Please end the sentences in the documentation with a period.

[1]: https://docs.rs/tokio/1/tokio/sync/struct.Mutex.html#method.lock

**Comment by ijackson** [comment]
> I think having a method like this makes sense. Some comments:
> 
>  1. I prefer to add the method to all `RwLock` guards, or to none. It should be possible to add it to the remaining guards by changing the `&'a Semaphore` field to `&'a RwLock`.

I tried this.  But the guard returned from `RwLock*Guard::map` needs to be `RwLockReadGuard<U>` but it needs to contain `&RwLock<T>`.  So it would have to have an additional generic parameter like `OwnedRwLock*Guard`.  Presumably this is why these types currently contain `&Semaphore`.  The other libraries that provide functions for recovering the mutex type have separate mapped guard types.

Is it OK to add that new generic (with the default, like for `OwnedRwLock*Guard`)?  I think that would be a breaking change, because currently a program might use `RwLock*Guard`s with different `T`s ,but the same `U`, interchangeably.  Currently they're the same type, but they would become different types.

>  2. We should not call the method `lock()`. That is confusing since we already have a [`Mutex::lock`](https://docs.rs/tokio/1/tokio/sync/struct.Mutex.html#method.lock) method, and that method does something different.

Good point.

I will address your other comments.  I'm inclined to split the `RwLock` versions into a separate MR to unblock the ones on `Mutex`.

**Comment by ijackson** [comment]
> > 1. We should not call the method `lock()`. That is confusing since we already have a [`Mutex::lock`](https://docs.rs/tokio/1/tokio/sync/struct.Mutex.html#method.lock) method, and that method does something different.
> 
> Good point.

Oh: another possible colour for this bikeshed.  `smol` calls this function `source`.

**Comment by ijackson** [comment]
>  1. I prefer to add the method to all `RwLock` guards, or to none. It should be possible to add it to the remaining guards by changing the `&'a Semaphore` field to `&'a RwLock`.

I split this off - see #3930.  Thanks.

>  2. We should not call the method `lock()`. That is confusing since we already have a [`Mutex::lock`](https://docs.rs/tokio/1/tokio/sync/struct.Mutex.html#method.lock) method, and that method does something different.

I was wrong when I said that `smol` uses `lock()` for this; it uses `source()`.  Let me know what name you prefer and I will paint the bikeshed accordingly.

>  3. Using `#[tokio::main] async fn main(){...}` for the doctests is fine, though all the other doctests put the example directly in `main` rather than define a new method. I think I'm ok either way, but putting it directly in `main` reduces the amount of hidden code.

Thanks.  I prefer to have a self-contained function as it avoids the example being cluttered with setup code, so I have left this as-is.

>  4. I'd like to include the imports in the doctests. Without them, the example would not compile if pasted into the playground.
>  5. Please run the doctests through rustfmt for more consistent formatting with the rest of the tests.
>  6. Please end the sentences in the documentation with a period.

Done.

Is there some automatic way to run rustfmt on doctests?  I resorted to cut-and-paste plus this rune: `perl -pe 's{^ */// ?(\# ?)?}{}' <t.rs >u.rs && cp u.rs v.rs && rustfmt --edition=2018 v.rs && diff -u u.rs v.rs`

**Comment by Darksonn** [comment]
Eh, I prefer `fn mutex()` over `fn source()`.

**Comment by ijackson** [comment]
> Eh, I prefer `fn mutex()` over `fn source()`.

Cool, no repainting needed then.  Anything else you'd like me to do to this MR?

**Comment by Darksonn** [comment]
I think it seems ok. I'll wait for the americans to get a chance to review it though.

**Review by Darksonn — APPROVED** [review]

**Pull request merged**
